### PR TITLE
Add live tests to release, main and new live-test workflows

### PR DESCRIPTION
**Pull request by hallipr**

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -5,8 +5,10 @@ on:
     types: [published]
 
 jobs:
-  deploy:
+  build:
+    name: Build
     runs-on: ubuntu-latest
+    environment: Live Testing
```
Comment (weshaggard) [inline .github/workflows/release.yml:18]: Why not split this into build and test as well?

File: .github/workflows/live-test.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+name: Live Test
```
Comment (weshaggard) [inline .github/workflows/live-test.yml:2]: Is this one needed to be separate? Can we not just use the main.yml with a manual or scheduled trigger?

Comment (hallipr) [reply .github/workflows/live-test.yml:2]: Main runs build, unit test and deploy.   We will have already run these for any PR.   We're only lacking the ability to run live tests against the PR.

File: .github/workflows/release.yml (hunk 1 above)
Comment (hallipr) [reply .github/workflows/release.yml:18]: Release will depend on live tests and live test should only run if unit tests have passed. As there's no parallel work, I didn't see the need to run this in a separate job. 

The same could be said for the Deploy job.   I don't see any reason for it to be its own job.

Comment (hallipr) [reply .github/workflows/release.yml:18]: ... except for retries.   I would be handy to be able to retry a deployment after a long live test run

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -1,7 +1,14 @@
-name: Build and Test
+# This worflow is triggered on push to main, pull request, or by manual invocation.
+# It builds and unit tests the codebase.
+name: Build and Unit Test
 
 on:
   workflow_dispatch:
+    inputs:
```
Comment (weshaggard) [inline .github/workflows/main.yml]: I don't think this is needed given they can trigger against a PR already and they can trigger against a branch as needed.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -20,34 +31,69 @@ jobs:
...
+      contents: write
+    env:
+      NUGET_API_KEY: ${{ github.event_name == 'release' && secrets.NUGET_API_KEY || secrets.GITHUB_TOKEN }}
+      PUBLISH_TARGET: ${{ github.event_name == 'release' && 'https://api.nuget.org/v3/index.json' || 'github' }}
+    steps:
+      - uses: actions/download-artifact@v2
+
       # Append the nuget package to the github release that triggered this workflow
       - name: Upload release asset
+        if: github.event_name == 'release'
```
Comment (weshaggard) [inline .github/workflows/release.yml:78]: I think we always want to publish the artifact.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -20,34 +31,69 @@ jobs:
...
+        uses: actions/upload-artifact@v2
+        if: ${{ !cancelled() }}
+        with:
+          name: build-artifacts
+          path: ${{ github.workspace }}/artifacts
+
+  deploy:
+    name: Publish Package
+    needs: build
+    if: github.event_name != 'pull_request'
```
Comment (weshaggard) [inline .github/workflows/release.yml]: is it even possible for this to be triggered from a PR?

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -20,34 +31,69 @@ jobs:
...
+        run: dotnet nuget add source
+          "https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json"
+          --name "github"
+          --username ${{ github.actor }}
+          --password ${{ secrets.GITHUB_TOKEN }}
+          --store-password-in-clear-text
 
-      - name: Publish
+      - name: Publish NuGet package
         run: dotnet nuget push
```
Comment (weshaggard) [inline .github/workflows/release.yml:100]: This needs a condition of ==release

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -20,34 +31,69 @@ jobs:
...
+    name: Publish Package
+    needs: build
+    if: github.event_name != 'pull_request'
+    runs-on: ubuntu-latest
+    permissions:
+      packages: write
+      contents: write
+    env:
+      NUGET_API_KEY: ${{ github.event_name == 'release' && secrets.NUGET_API_KEY || secrets.GITHUB_TOKEN }}
+      PUBLISH_TARGET: ${{ github.event_name == 'release' && 'https://api.nuget.org/v3/index.json' || 'github' }}
```
Comment (weshaggard) [inline .github/workflows/release.yml]: Where is this used?

File: .github/workflows/release.yml (hunk 4 above)
Comment (hallipr) [reply .github/workflows/release.yml:78]: This is attaching the artifacts to the Release resource.   Build artifacts are always upload as the last step in the build job

File: .github/workflows/release.yml (hunk 5 above)
Comment (hallipr) [reply .github/workflows/release.yml]: condition removed

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -1,53 +1,101 @@
...
 
-      - name: Publish
+      - name: Publish NuGet package
         run: dotnet nuget push
-          ${{github.workspace}}/artifacts/packages/*.nupkg
-          --source https://api.nuget.org/v3/index.json
-          --api-key ${{ secrets.NUGET_API_KEY }}
-          --skip-duplicate
+            ${{github.workspace}}/build-artifacts/packages/*.nupkg
+            --source ${{ env.NUGET_API_KEY }}
```
Comment (weshaggard) [inline .github/workflows/release.yml]: Should this be `PUBLISH_TARGET`?

File: .github/workflows/release.yml (hunk 6 above)
Comment (hallipr) [reply .github/workflows/release.yml:100]: We publish the scheduled runs to the local feed and release to nuget.org.   The conditionals in the job variables pick the feed and token to use

File: .github/workflows/release.yml (hunk 7 above)
Comment (hallipr) [reply .github/workflows/release.yml]: in the Publish NuGet package step.   I copy pasted wrong before, but it's corrected now

File: .github/workflows/release.yml (hunk 8 above)
Comment (hallipr) [reply .github/workflows/release.yml]: corrected

File: .github/workflows/main.yml (hunk 3 above)
Comment (hallipr) [reply .github/workflows/main.yml]: removed

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -1,53 +1,104 @@
...
+          ${{ github.workspace }}/build-artifacts/packages/*.*nupkg
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
-      - name: Upload artifact
-        uses: actions/upload-artifact@v2
-        with:
-          name: build-artifacts
-          path: ${{github.workspace}}/artifacts
+      - name: NuGet autenticate
```
Comment (weshaggard) [inline .github/workflows/release.yml]:
```suggestion
      - name: NuGet authenticate
```

File: .github/workflows/live-test.yml [hunk 10]
```diff
@@ -0,0 +1,45 @@
+# This workflow is triggered by the user and runs live tests on the codebase.
+name: Live Test
+
+on:
+  workflow_dispatch:
```
Comment (weshaggard) [inline .github/workflows/live-test.yml:5]: Didn't we decide to switch this trigger to be label based?

File: .github/workflows/live-test.yml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+jobs:
+  test:
+    name: Live Test
+    runs-on: ubuntu-latest
+    environment: Live Testing
+    env:
+      version_suffix_args: ${{ format('--version-suffix="alpha.{0}"', github.run_number) }}
+    steps:
+      - name: Setup .NET
+        uses: actions/setup-dotnet@v1
```
Comment (weshaggard) [inline .github/workflows/live-test.yml]:
```suggestion
        uses: actions/setup-dotnet@v3
```

**Review by weshaggard — APPROVED** [review]
One last question about the live-test trigger but otherwise looks good.

File: .github/workflows/live-test.yml (hunk 10 above)
Comment (hallipr) [reply .github/workflows/live-test.yml:5]: I was just about to push that change :)

**Pull request merged**